docs/adr/0002-resource-definition.md

```diff
@@ -0,0 +1,149 @@
+# Resource definition
+
+* Status: proposed
+* Deciders: @dunglas @soyuka @vincentchalamon @GregoireHebert
+
+## Context and Problem Statement
+
+The API Platform `@ApiResource` annotation was initially created to represent a Resource as defined in [Roy Fiedling's dissertation about REST](https://www.ics.uci.edu/~fielding/pubs/dissertation/rest_arch_style.htm#sec_5_2_1_1) in corelation with [RFC 7231 about HTTP Semantics](https://httpwg.org/specs/rfc7231.html#resources). This annotation brings some confusion as it mixes concepts of resources and operations. Here we discussed how we could revamp API Platform's resource definition using PHP8 attributes, beeing as close as we can to Roy Fiedling's thesis vocabulary.
+
+## Considered Options
+
+* Declare multiple ApiResource on a PHP Class [see Subresources definition](./0000-subresources-definition.md)
+* Declare operations in conjunction with resources using two attributes: `Resource` and `Operation`
+* Use HTTP Verb to represent operations with a syntax sugar for collections (`CGET`?)
+
+## Decision Outcome
+
+As Roy Fiedling's thesis states:
+
+> REST uses a resource identifier to identify the particular resource involved in an interaction between components. REST connectors provide a generic interface for accessing and manipulating the value set of a resource, regardless of how the membership function is defined or the type of software that is handling the request. 
+
+In API Platform, this resource identifier is also named [IRI (Internationalized Resource Identifiers)](https://tools.ietf.org/html/rfc3987). Following these recommandation, applied to PHP, we came up with the following attributes:
+
+```php
+<?php
+
+#[Get]
+#[Post]
+class Users
+{
+}
+
+#[Get]
+#[Put]
+#[Delete]
+class User
+{
+    public $id;
+}
+```
+
+Under the hood, API Platform would declare two routes Representing the `/users` resource:
+
+- GET /users
+- POST /users
+
+and three routes representing the `/users/{id}` resource:
+
+- GET /users/{id}
+- PUT /users/{id}
+- DELETE /users/{id}
+
+For convenience and to ease the upgrade path, these would still be available on a single class:
+
+```php
+<?php
+
+#[Resource]
+class User {}
+```
+
+To replace some operations on this resource, we would specify HTTP verbs to override:
+
+```php
+<?php
+
+#[Resource]
+#[Get]
+class User {}
+```
+
+would declare:
+
+- GET /users/{id}
+
+For collections:
+
+```php
+<?php
+
+#[Resource]
+#[CGet]
```

Comment:
What about going a bit more descriptive way: `#[List]` ?

Reply:
List would only work for `get`, here collection would be used to declare routes without an identifier for example `POST /users` and `GET /users`.

Reply:
Hmmm I know it's not common, but I would personally do it like:
```php
#[List] <-- GET /users
#[Create] <-- POST /users